lucene/analysis/smartcn/src/java/org/apache/lucene/analysis/cn/smart/hhmm/WordDictionary.java

```diff
@@ -0,0 +1,526 @@
+          wordIndexTable[i] = -1;
+        }
+        wordItem_charArrayTable = new char[GB2312_CHAR_NUM][][];
+        wordItem_frequencyTable = new int[GB2312_CHAR_NUM][];
+        // int total =
+        loadMainDataFromFile(dctFilePath);
+        expandDelimiterData();
+        mergeSameWords();
+        sortEachItems();
+        // log.info("load dictionary: " + dctFilePath + " total:" + total);
+      } catch (IOException e) {
+        throw new RuntimeException(e.getMessage());
+      }
+
+      saveToObj(serialObj);
+    }
+
+  }
+
+  /**
+   * Load coredict.mem internally from the jar file.
+   * 
+   * @throws IOException If there is a low-level I/O error.
+   */
+  public void load() throws IOException, ClassNotFoundException {
+    InputStream input = this.getClass().getResourceAsStream("coredict.mem");
+    loadFromObjectInputStream(input);
+  }
+
+  private boolean loadFromObj(Path serialObj) {
+    try {
+      loadFromObjectInputStream(Files.newInputStream(serialObj));
+      return true;
+    } catch (Exception e) {
+      throw new RuntimeException(e);
+    }
+  }
+
+  @SuppressForbidden(reason = "TODO: fix code to serialize its own dictionary vs. a binary blob in the codebase")
+  private void loadFromObjectInputStream(InputStream serialObjectInputStream)
+      throws IOException, ClassNotFoundException {
+    try (ObjectInputStream input = new ObjectInputStream(serialObjectInputStream)) {
+      wordIndexTable = (short[]) input.readObject();
+      charIndexTable = (char[]) input.readObject();
+      wordItem_charArrayTable = (char[][][]) input.readObject();
+      wordItem_frequencyTable = (int[][]) input.readObject();
+      // log.info("load core dict from serialization.");
+    }
+  }
+
+  @SuppressForbidden(reason = "TODO: fix code to serialize its own dictionary vs. a binary blob in the codebase")
+  private void saveToObj(Path serialObj) {
+    try (ObjectOutputStream output = new ObjectOutputStream(Files.newOutputStream(serialObj))) {
+      output.writeObject(wordIndexTable);
+      output.writeObject(charIndexTable);
+      output.writeObject(wordItem_charArrayTable);
+      output.writeObject(wordItem_frequencyTable);
+      // log.info("serialize core dict.");
+    } catch (Exception e) {
+      // log.warn(e.getMessage());
+    }
+  }
+
+  /**
+   * Load the datafile into this WordDictionary
+   * 
+   * @param dctFilePath path to word dictionary (coredict.dct)
+   * @return number of words read
+   * @throws IOException If there is a low-level I/O error.
+   */
+  private int loadMainDataFromFile(String dctFilePath) throws IOException {
+    int i, cnt, length, total = 0;
+    // The file only counted 6763 Chinese characters plus 5 reserved slots 3756~3760.
+    // The 3756th is used (as a header) to store information.
+    int[] buffer = new int[3];
+    byte[] intBuffer = new byte[4];
+    String tmpword;
+    DataInputStream dctFile = new DataInputStream(Files.newInputStream(Paths.get(dctFilePath)));
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code contains a resource that might not be closed properly. Resource leaks can cause your system to slow down or crash.

**Fix**
Consider closing the following resource: *dctFile*. The resource is referenced in statements at the following lines: 189, 196, 210, 213, 216, 226. The resource closure statement is at line: 242. There are other execution paths that do not contain closure statements, e.g., when exception is thrown by DataInputStream.read. Either a) close *dctFile* in a try-finally block or b) close the resource by declaring *dctFile* in a try-with-resources block.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).